### ISSUE TITLE
Adding an option to return residual in jax.nn.dot_product_attention

### ISSUE THREAD

**Issue by young-geng** [jax-ml/jax#32484]
Currently, the `jax.nn.dot_product_attention` function only returns the attention output. It would be very useful to have an option to return also the logsumexp residual of the softmax, similar to how it is implemented in [`jax._src.cudnn.fused_attention_stablehlo.dot_product_attention`](https://github.com/jax-ml/jax/blob/2892bf1e1ccd60d01bb818ec7f888d2fe05c4aca/jax/_src/cudnn/fused_attention_stablehlo.py#L1924). The API would look this this:

```python
output, lse = jax.nn.dot_product_attention(query, key, bias, return_residual=True)
```

This is very useful when implementing sharded attention (ring attention, zigzag ring attention) where the final attention is calculated in chunks of key and value. Since the CUDNN backend already supports this, it shouldn't be difficult to add to `jax.nn.dot_product_attention`, because the XLA backend is just implemented in pure jax. Here's an example of using the `return_residual` option to implement ring attention:

```python
from functools import partial

import numpy as np
import jax
import jax.numpy as jnp
from jax.experimental import shard_map
from jax.sharding import Mesh, PartitionSpec as P
import einops

from jax._src.cudnn.fused_attention_stablehlo import dot_product_attention as cudnn_dot_product_attention

def cudnn_ring_attention(q, k, v, mesh, axis_name, causal=True):
    sequence_sharding = P(None, axis_name, None, None)
    @partial(
        shard_map.shard_map,
        mesh=mesh,
        in_specs=(sequence_sharding, sequence_sharding, sequence_sharding),
        out_specs=sequence_sharding,
        check_rep=False,
    )
    def apply_cudnn_ring_attention(q, k, v):
        axis_size = jax.lax.axis_size(axis_name)
        axis_index = jax.lax.axis_index(axis_name)
        shard_size = q.shape[1]

        positions = jnp.arange(shard_size, dtype=jnp.int32)
        q_positions = positions.reshape(-1, 1) + axis_index * shard_size

        def process_kv_shard(ring_iteration, carry):
            partial_attention_state, k, v = carry
            normalized_partial_softmax, partial_logsumexp = partial_attention_state

            kv_index = (axis_index - ring_iteration) % axis_size
            kv_positions = positions.reshape(1, -1) + kv_index * shard_size
            mask = einops.rearrange(
                q_positions >= kv_positions,
                'q k -> 1 1 q k'
            )

            flash_attention_output, residual = cudnn_dot_product_attention(
                query=q, key=k, value=v,
                mask=mask if causal else None,
                scale=1.0 / np.sqrt(q.shape[-1]),
                qkv_layout='BTNH',
                return_residual=True,
            )
            residual = einops.rearrange(residual, 'b h s -> b s h 1')
            new_logsumexp = jnp.logaddexp(partial_logsumexp, residual)
            w1 = jnp.exp(partial_logsumexp - new_logsumexp)
            w2 = jnp.exp(residual - new_logsumexp)
            new_partial_softmax = (
                normalized_partial_softmax * w1 + flash_attention_output * w2
            )
            partial_attention_state = (new_partial_softmax, new_logsumexp)

            k, v = jax.lax.ppermute(
                (k, v),
                axis_name=axis_name,
                perm=[(i, (i + 1) % axis_size) for i in range(axis_size)]
            )
            return partial_attention_state, k, v

        init_partial_attention_state = (
            jnp.zeros_like(q, dtype=jnp.float32),
            -jnp.inf * jnp.ones((*q.shape[:-1], 1), dtype=jnp.float32),
        )

        final_softmax, _ = jax.lax.fori_loop(
            0, axis_size, process_kv_shard, (init_partial_attention_state, k, v)
        )[0]

        return final_softmax

    return apply_cudnn_ring_attention(q, k, v)
```

**Comment by nadtech-hub** [comment]
May I take on it?

**Comment by young-geng** [comment]
Implemented in [this PR](https://github.com/jax-ml/jax/pull/32657).
